#  Dataviz project
## ARTAUD Lucas & SIVASUBRAMANIAM Iswarya DIA 1

In [184]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import geopandas as gpd
import plotly.express as px
import pycountry_convert as pc 
import plotly.graph_objects as go
import plotly.subplots as sp
from geopy.geocoders import Nominatim
import numpy as np


## Dataset

In [185]:
df=pd.read_csv("Life Expectancy Data.csv")

In [186]:
df

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
2933             27     4.36                0.000000         68.0        31   
2934             26     4.06                0.000000          7.0       998   
2935             25     4.43                0.000000         73.0       304   
2936             25     1.72                0.000000         76.0       529   
2937             24     1.68                0.000000         79.0      1483   

      ...  Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  \
0     ...    6.0               8.16         65.0        0.1  584.259210   
1     ...   58.0               8.18         62.0        0.1  612.696514   
2     ...   62.0               8.13         64.0        0.1  631.744976   
3     ...   67.0               8.52         67.0        0.1  669.959000   
4     ...   68.0               7.87         68.0        0.1   63.537231   
...   ...    ...                ...          ...        ...         ...   
2933  ...   67.0               7.13         65.0       33.6  454.366654   
2934  ...    7.0               6.52         68.0       36.7  453.351155   
2935  ...   73.0               6.53         71.0       39.8   57.348340   
2936  ...   76.0               6.16         75.0       42.1  548.587312   
2937  ...   78.0               7.10         78.0       43.5  547.358878   

      Population   thinness  1-19 years   thinness 5-9 years  \
0     33736494.0                   17.2                 17.3   
1       327582.0                   17.5                 17.5   
2     31731688.0                   17.7                 17.7   
3      3696958.0                   17.9                 18.0   
4      2978599.0                   18.2                 18.2   
...          ...                    ...                  ...   
2933  12777511.0                    9.4                  9.4   
2934  12633897.0                    9.8                  9.9   
2935    125525.0                    1.2                  1.3   
2936  12366165.0                    1.6                  1.7   
2937  12222251.0                   11.0                 11.2   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427      

In [187]:
df.shape

(2938, 22)

In [188]:
df.isna().sum()

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

In [189]:
df.dropna(inplace=True)


## Context and motivation

##### The data set chosen is a statistical data on factors influencing Life Expectancy. The data comes from the World Helath Organization for a period of time of 15 years.   

##### Many studies in the past have explored the factors influencing life expectancy, centering around demographic variables, income composition, and mortality rates. However, these studies often neglected the impact of immunization and the Human Development Index. Additionally, some prior research relied on a one-year dataset for all countries but doing this research r a period of time of 15 year enables us to visualise the changes over time. This data set allows us to do a country based observation to identify the main factors that are contributing to lower life expectancy.  
##### This data-set counts health factors of 193 countries over 15 years (2000-2015). We have 22 columns and 2938 rows (befor cleaning).

##### From our preliminary analysis, population, Hepatitis B and GDP columns contains most of the missing data.We decided to remove all the missing values in order to have a clean data.

##### The motivation of our project is to analyze the different factors taht influences the life expectancy and compare on different scales (between continents, developed / developing etc...).
##### We hope to learn more about the various factors affecting life expectancy and how those findings can guide public health interventions and policies. Targeted healthcare initiatives could be guided, for instance, by identifying particular regions or demographic groups experiencing challenges with life expectancy. Furthermore, knowledge of how social determinants, economic variables, and immunizations affect life expectancy can support evidence-based decision-making at the national and international levels. 

#### Columns

* **Country**: Country
* **Year**: Year
* **Status**: Classification of countries as 'developed' or 'developing' based on their gross domestic product(GDP).
* **Life expectancy**: Life expectancy (years of age)
* **Adult Mortality**: Adult Mortality Rates of both sexes (Probability of dying between 15 and 60 years per 1000 population)
* **infant deaths**: Number of Infant (0-1 year of age) Deaths per 1000 population.
* **Alcohol**: Alcohol, recorded per capita (15+) consumption (in litres of pure alcohol).
* **Percentage expenditure**: Expenditure on health as a percentage of GPD per capita. (%)
* **Hepatitis B**: Hepatitis B immunization coverage among 1-year-olds. (%)
* **Measles**: Number of reported cases per 1000 population.
* **BMI**: Average Body Mass index of entire population
* **Under-five deaths**: Number of under-five deaths per 1000 population
* **Polio**: Polio immunization coverage among 1-year-olds (%)
* **Total expenditure**: General government expenditure on health as a percentage of total government expenditure (%)
* **Diphtheria**: Diphtheria tetanus toxoid and pertussis (DTP3) immunization coverage among 1-year-olds (%)
* **HIV/AIDS**: Deaths per 1000 live births HIV/AIDS (0-4 years)
* **GDP**: Gross Domestic Product per capita (in USD)
* **Population**: Population of the country
* **Thinness 1-19 years**: Prevalence of thinness among children and adolescents for Age 10 to 19 (%) 
* **Thinness 5-9 years**: Prevalence of thinness among children for Age 5 to 9 (%)
* **Income composition of resources**: Human Development Index in terms of income composition of resources (index ranging from 0 to 1)
* **Schooling**: Number of years of Schooling (years)

## Visualisation

### Creation of a new column continent

With the library pycountry_convert we are going to create a new column "Continent" that will correspond to the continent of the country.

In [190]:
continent_name = {
    'AF': 'Africa',
    'AS': 'Asia',
    'EU': 'Europe',
    'NA': 'North America',
    'OC': 'Oceania',
    'SA': 'South America',
    'AN': 'Antarctica'
}

In [191]:
# Replacing the special cases
df['Country'] = df['Country'].replace({'Bolivia (Plurinational State of)': 'Bolivia'})

In [192]:
def convert(row):
    # convert country name to country code
    country_code =pc.country_name_to_country_alpha2(row.Country,cn_name_format="default")
    # convert country_code to continent code
    try:
        continent_code = pc.country_alpha2_to_continent_code(country_code)
        return continent_name.get(continent_code, None)
    except :
        return None

In [193]:
df['Continent']= df.apply(convert,axis=1)

In [194]:
df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling', 'Continent'],
      dtype='object')

### Global analysis

#### 1. Average life expectancy and population over the years

In [195]:
average_life_expectancy_yearly = df.groupby('Year')['Life expectancy '].mean().reset_index()
average_population_yearly = df.groupby('Year')['Population'].mean().reset_index()

In [196]:
fig = sp.make_subplots(rows=2, cols=1, subplot_titles=['Average Life Expectancy', 'Average Population'])

fig.add_trace(
    go.Scatter(x=average_life_expectancy_yearly['Year'], y=average_life_expectancy_yearly['Life expectancy '],
               mode='lines', name='Life Expectancy'),
    row=1, col=1
)

fig.add_trace(
    go.Scatter(x=average_population_yearly['Year'], y=average_population_yearly['Population'],
               mode='lines', name='Population'),
    row=2, col=1
)

fig.update_layout(
    font=dict(size=12),
    width=1500,
    height=600,
    margin=dict(l=20, r=20, t=60, b=20),
)
fig.show()

#### 2. Life Expectancy over the years of the top 10 and bottom 10 countries

In [197]:
average_life_expectancy = df.groupby('Country')['Life expectancy '].mean().reset_index() # calcultae the average life expectancy
top10_countries = average_life_expectancy.nlargest(10, 'Life expectancy ') # take the top 10 average life expectancy
bottom10_countries = average_life_expectancy.nsmallest(10, 'Life expectancy ') # take the bottom 10 average life expectancy

In [198]:
# Filtering the Original DataFrame for the Selected Countries
selected_countries = top10_countries['Country'].tolist() + bottom10_countries['Country'].tolist()
filtered_df = df[df['Country'].isin(selected_countries)] 

In [199]:
selected_countries

['Ireland',
 'Canada',
 'France',
 'Italy',
 'Spain',
 'Australia',
 'Sweden',
 'Austria',
 'Netherlands',
 'Greece',
 'Sierra Leone',
 'Lesotho',
 'Zimbabwe',
 'Malawi',
 'Angola',
 'Swaziland',
 'Central African Republic',
 'Chad',
 'Nigeria',
 'Mozambique']

In [200]:
fig = px.line(filtered_df, x='Year', y='Life expectancy ', color='Country',title='Life Expectancy Over the Years for the top 5 and bottom 5 Selected Countries')
fig.update_layout(
    xaxis_title='Year',
    yaxis_title='Life Expectancy',
    legend_title='Country',
    font=dict(size=15),
    width=1500,
    height=1000,
    margin=dict(l=20, r=20, t=60, b=20))

fig.show()

#### 3.Violin Plot for Life Expectancy by Continent 

In [201]:
fig = px.violin(df, x='Continent', y='Life expectancy ', color='Continent',box=True,title='Violin Plot for Life Expectancy and Population by Continent')
 
fig.update_layout(
    xaxis_title='Continent',
    yaxis_title='Life Expectancy',
    legend_title='Continent',
    font=dict(size=12),
    width=1500,
    height=1000,
    margin=dict(l=20, r=20, t=60, b=20),
)

fig.show()

#### 4. Pie chart for the distribution of countries by Status

In [202]:

fig = px.pie(
    df, 
    names='Status', 
    title='Distribution of Countries by Development Status',
    color='Status',  # Assigning colors based on the 'Status' column
    color_discrete_sequence=['#67001F', '#F4A582'],
    hole=0.4, 
)

# Update layout for better readability
fig.update_layout(
    width=1500,
    height=1000,
    legend_title_text='Development Status',
)

fig.show()

#### 5. Comparing the GDP from 2000 to 2015 by the status


In [203]:
df_filtered = df[df['Year'].between(2000, 2014)]  # Exclude the year 2015 becaus it is not complete
df_gdp_avg=df_filtered.groupby(['Year', 'Status'])['GDP'].mean().reset_index()


In [204]:
fig = px.bar(df_gdp_avg, x='Status', y='GDP', color='Status', title='Average GDP by Country Status Over Years',animation_frame='Year', barmode='group')
fig.update_layout(xaxis_title='Year', yaxis_title='Average GDP (USD)', width=1200, height=800)
fig.show()

#### 6. Comparing the life expectancy from 2000 to 2015 by the status

In [205]:
fig = px.box(df_filtered, x='Status', y='Life expectancy ',title='Life Expectancy Distribution by Status (2000 to 2014)',animation_frame='Year',labels={'Life expectancy': 'Life Expectancy', 'Status': 'Development Status'})

fig.update_layout(
    xaxis_title='Development Status',
    yaxis_title='Life Expectancy',
    width=1200,
    height=1200
)
fig.frames = fig.frames[::-1] # reversing the frame in order to have a progessive plot
fig.show()

#### 7.Correlation map in order to study the columns that are influencing the life expectancy

In [206]:
numeric_columns = df.select_dtypes(include=['float64']).columns # Filtering the non-numeric columns in order to do a correlation
correlation_matrix = df[numeric_columns].corr()

In [207]:
fig = px.imshow(
    correlation_matrix,
    color_continuous_scale='GnBu',
    labels=dict(x='Features', y='Features', color='Correlation'),
    title='Correlation Heatmap',
    width=1500,
    height=1000,
)
fig.show()

Let's concentrate on the analysis of schooling, income composition of ressources and BMI.

#### 8. Correlation between schooling and life expectancy by continent in 2014

In [208]:
fig = px.scatter(df, x='Schooling', y='Life expectancy ', trendline="ols",title='Scatter Plot of Schooling vs Life Expectancy',labels={'Schooling': 'Years of Schooling', 'Life expectancy ': 'Life Expectancy'})

fig.update_layout(
    xaxis_title='Years of Schooling',
    yaxis_title='Life Expectancy',
    width=1800,
    height=1000,
)
fig.update_traces(
    line=dict(color='red', dash='solid'), 
    selector=dict(mode='lines')
)
fig.show()

#### 9. Correlation between income ressources and life expectancy by continent in 2014

In [209]:
df_2014 = df[df['Year'] == 2014]

In [210]:
fig = px.scatter(df_2014, x='Income composition of resources', y='Life expectancy ', color='Continent',hover_name='Country', title='Income Composition vs Life Expectancy',labels={'Income composition of resources': 'Income Composition of Resources', 'Life expectancy': 'Life Expectancy'})

fig.update_layout(
    xaxis_title='Income Composition of Resources',
    yaxis_title='Life Expectancy',
    width=1800,
    height=1000,
)
fig.update_traces(marker=dict(size=8))

fig.show()

#### 10.Average BMI by continent

In [211]:
df_avg_bmi = df[df['Year'].between(2000, 2014)].groupby(['Continent', 'Year'])[' BMI '].mean().reset_index().sort_values(by=' BMI ', ascending=False)


In [212]:
fig = px.bar(df_avg_bmi, x='Continent', y=' BMI ', color='Continent',animation_frame='Year',title='Average BMI Over the Years by Continent (2000-2014)',labels={' BMI ': 'Average BMI', 'Continent': 'Continent'},range_y=[df_avg_bmi[' BMI '].min(), df_avg_bmi[' BMI '].max()])
fig.update_layout(
    xaxis_title='Continent',
    yaxis_title='Average BMI',
    width=1800,
    height=1000,
)
fig.frames = fig.frames[::-1] # reversing the frame in order to have a progessive plot>
fig.show()

#### 11.Comparision on the evolution of HIV and Measles

In [213]:
df_hiv_measles = df.groupby('Year')[[' HIV/AIDS', 'Measles ']].sum().reset_index()

In [214]:
fig = sp.make_subplots(rows=2, cols=1, subplot_titles=['HIV/AIDS', 'Measles'])

bar_trace_hiv = go.Bar(x=df_hiv_measles['Year'], y=df_hiv_measles[' HIV/AIDS'], name='HIV/AIDS')
bar_trace_measles = go.Bar(x=df_hiv_measles['Year'], y=df_hiv_measles['Measles '], name='Measles')

fig.add_trace(bar_trace_hiv, row=2, col=1)
fig.add_trace(bar_trace_measles, row=1, col=1)

fig.update_layout(
    title='Number of Deaths by HIV and Measles (2000-2015)',
    xaxis_title='Year',
    yaxis_title='Number of Deaths',
    width=1800,
    height=1000,
)
fig.update_xaxes(tickmode='linear')
fig.update_xaxes( title_text='Year', row=2, col=1)

fig.update_yaxes( title_text='Number of Deaths')
fig.show()


#### 12. Map on the evolution of Adult Mortality 

In [215]:
unique_countries = pd.DataFrame(df['Country'].unique(), columns=['Country'])
geolocator = Nominatim(user_agent="my_geocoder")


In [216]:
def get_lat_lon(country):
    location = geolocator.geocode(country)
    if location:
        return pd.Series({'latitude': location.latitude, 'longitude': location.longitude})
    else:
        return pd.Series({'latitude': None, 'longitude': None})

In [217]:
unique_countries[['latitude', 'longitude']] = unique_countries['Country'].apply(get_lat_lon)

In [218]:
unique_countries

Country   latitude   longitude
0    Afghanistan  33.768006   66.238514
1        Albania   1.231526  -75.892043
2        Algeria  28.000027    2.999983
3         Angola -11.877577   17.569124
4      Argentina -34.996496  -64.967282
..           ...        ...         ...
128      Uruguay -32.875555  -56.020153
129   Uzbekistan  41.323730   63.952810
130      Vanuatu -16.525507  168.106915
131       Zambia -14.518912   27.558988
132     Zimbabwe -18.455496   29.746841

[133 rows x 3 columns]

In [219]:
df = pd.merge(df, unique_countries, on='Country', how='left')

In [220]:
df

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
1644     Zimbabwe  2004  Developing              44.3            723.0   
1645     Zimbabwe  2003  Developing              44.5            715.0   
1646     Zimbabwe  2002  Developing              44.8             73.0   
1647     Zimbabwe  2001  Developing              45.3            686.0   
1648     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01               71.279624         65.0      1154   
1                64     0.01               73.523582         62.0       492   
2                66     0.01               73.219243         64.0       430   
3                69     0.01               78.184215         67.0      2787   
4                71     0.01                7.097109         68.0      3013   
...             ...      ...                     ...          ...       ...   
1644             27     4.36                0.000000         68.0        31   
1645             26     4.06                0.000000          7.0       998   
1646             25     4.43                0.000000         73.0       304   
1647             25     1.72                0.000000         76.0       529   
1648             24     1.68                0.000000         79.0      1483   

      ...   HIV/AIDS         GDP  Population   thinness  1-19 years  \
0     ...        0.1  584.259210  33736494.0                   17.2   
1     ...        0.1  612.696514    327582.0                   17.5   
2     ...        0.1  631.744976  31731688.0                   17.7   
3     ...        0.1  669.959000   3696958.0                   17.9   
4     ...        0.1   63.537231   2978599.0                   18.2   
...   ...        ...         ...         ...                    ...   
1644  ...       33.6  454.366654  12777511.0                    9.4   
1645  ...       36.7  453.351155  12633897.0                    9.8   
1646  ...       39.8   57.348340    125525.0                    1.2   
1647  ...       42.1  548.587312  12366165.0                    1.6   
1648  ...       43.5  547.358878  12222251.0                   11.0   

       thinness 5-9 years  Income composition of resources  Schooling  \
0                    17.3                            0.479       10.1   
1                    17.5                            0.476       10.0   
2                    17.7                            0.470        9.9   
3                    18.0                            0.463        9.8   
4                    18.2                            0.454        9.5   
...                   ...                              ...        ...   
1644                  9.4                            0.407        9.2   
1645                  9.9                            0.418        9.5   
1646                  1.3                            0.427       10.0   
1647                  1.7                            0.427        9.8   
1648                 11.2                            0.434        9.8   

      Continent   latitude  longitude  
0          Asia  33.768006  66.238514  
1          Asia  33.768006  66.238514  
2          Asia  33.768006  66.238514  
3          Asia  33.768006  66.238514  
4          Asia  33.768006  66.238514  
...         ...        ...        ...  
1644     Africa -18.455496  29.746841  
1645     Africa -18.455496  29.746841  
1646     Africa -18.455496  29.746841  
1647     Africa -18.455496  29.746841  
16

In [221]:
fig = px.choropleth(
    df,
    locations='Country',
    locationmode='country names',
    color='Adult Mortality',
    hover_name='Country',
    color_continuous_scale=px.colors.sequential.Plasma,
    title='Adult Mortality Across Countries',
    template='plotly',
    animation_frame='Year'
)
fig.update_geos(
    resolution=110,
    showcoastlines=True,
    coastlinecolor="Black",
    showland=True,
    landcolor="white",
)
fig.update_layout(
    width=1500,  
    height=1000,  
)
fig.frames = fig.frames[::-1]
fig.show()

#### 13. Map on the evolution of under five death to compare

In [222]:
fig = px.choropleth(
    df,
    locations='Country',
    locationmode='country names',
    color='infant deaths',
    hover_name='Country',
    color_continuous_scale=px.colors.sequential.Plasma,
    title='Adult Mortality Across Countries',
    template='plotly',
    animation_frame='Year'
)
fig.update_geos(
    resolution=110,
    showcoastlines=True,
    coastlinecolor="Black",
    showland=True,
    landcolor="white",
)
fig.update_layout(
    width=1500,  
    height=1000,  
)
fig.frames = fig.frames[::-1]
fig.show()